### PR TITLE
Cleaned up some suspicious range checks

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
Steve and I noticed some seemingly useless range checks in UnmanagedMemoryStream. They were checking if long values were larger than Int64.MaxValue. I removed the checks and cleaned up related code a bit.

### CODE REVIEW

File: src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs [hunk 1]
```diff
@@ -429,34 +421,24 @@ public override long Position
...
 
-                // Note: subtracting pointers returns an Int64.  Working around
-                // to avoid hitting compiler warning CS0652 on this line. 
-                if (new IntPtr(value - _mem).ToInt64() > UnmanagedMemStreamMaxLength)
-                    throw new ArgumentOutOfRangeException("offset", Strings.ArgumentOutOfRange_UnmanagedMemStreamLength);
-                if (value < _mem)
-                    throw new IOException(Strings.IO_SeekBeforeBegin);
-
-                Interlocked.Exchange(ref _position, value - _mem);
+                Seek(value - _mem, SeekOrigin.Begin);
```
Comment (Petermarcu) [inline src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs]: Why this change? Doesn't it add some extra checks now on the path?

Comment (mellinoe) [inline src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs]: It does look like there is some duplicated argument validation for _isOpen and for the new seek position.

Comment (KrzysztofCwalina) [inline src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs]: I made the change, as I fundamentally think we should have a single routine that seeks the stream (per seek origin at least), otherwise it's very difficult to tell if they all do the same thing. Today, there are at lease three different routines that pretty much seek from the Begin origin.

This change does add some checks (and a switch statement). I considered restructuring the code the other way, i.e. Seek would call the set_Pointer method. I just assumed (possibly wrongly) that most people would use Seek instead of setting the position.

Longer term, I think we should fix code contracts, teach the compilers to elide them is repeated calls (like here), and then not worry too much about their impact.

Comment (Petermarcu) [inline src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs]: I agree with the spirit and longer term direction of converging on a single seek routine and relying on the compiler to reduce the contract checks. We don't have the compiler support today though. I don't feel strongly one way or the other and will leave it up to you.

Comment (stephentoub) [inline src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs]: Regarding duplication and differences, the Position property has additional logic applied for when IntPtr.Size == 4.  Do you know why that logic doesn't also exist in Seek?

Also, if I'm reading this correctly, I think your change introduces the possibility of an exception that wasn't previously there (this might be desirable, but I wanted to call it out, regardless).  PositionPointer could be set to a value such that value - _mem could be larger than Int64.MaxValue, and when cast to a long it'd end up wrapping around to be negative.  Previously that negative value would just be stored into _position in the Interocked.Exchange, but now it'll be passed to Seek, which will compare it against 0 and throw an exception for seeking before the beginning.

Comment (KrzysztofCwalina) [inline src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs]: I don't know why there is this difference, but I am glad you noticed it; this is exactly the problem I would like to avoid by having all seek-like logic in one place, including the preconditions for seeking.

As to the differences in error behavior (the place where exception is thrown), I would propose that we let this one go in the name of making seeking consistent, but I will chat with more people about it later today, as I do also worry about compat. I actually would like to do a major overhaul of this class. I really dislike how it's structured. Reasoning about this code, especially error code paths and boundary conditions is very hard. Part of the reason stems from the fact that it's two components mixed into one class (stream over byte* and stream over SafeBuffer), the way we do precondition checks, and the fact that we do unchecked arithmetic. But if I do such changes, it would be close to impossible for me to verify that error code paths stayed compatible.

Comment (stephentoub) [inline src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryStream.cs]: Thanks, @KrzysztofCwalina.  If after you speak with folks today the consensus is that the exceptional differences are ok, and if the changes to use Seek don't measurably and negatively affect perf, LGTM.

**Comment by KrzysztofCwalina** [comment]
We talked about it. I will try to refactor this whole component, submit a PR, and we will decide if the changes are too scary. 
I want to separate the two modes (the byte* and the SafeBuffer)
I want to cleanup all precondition checks to make them consistent and easy to reason about
I want to beef up tests to test for error code paths that we consider part of the contract

**Comment by Petermarcu** [comment]
@KrzysztofCwalina, can we close this PR as you prepare your alternative changes?

**Comment by mmitche** [comment]
test this please

**Comment by KrzysztofCwalina** [comment]
test this please

**Comment by KrzysztofCwalina** [comment]
Ok, I resubmitted the PR with a simple change that just removes the unnecessary checks. I will prototype the larger change separately. Could somebody CR this PR and comment or merge this change? Thanks.

**Comment by Petermarcu** [comment]
LGTM
